src/main/java/hudson/plugins/jira/Updater.java

```diff
@@ -148,55 +150,51 @@ boolean perform(Run<?, ?> build, TaskListener listener, AbstractIssueSelector se
             boolean useWikiStyleComments, boolean recordScmChanges, String groupVisibility, String roleVisibility) throws RestClientException {
 
         // copy to prevent ConcurrentModificationException
-        List<JiraIssue> copy = new ArrayList<JiraIssue>(issues);
+        Set<JiraIssue> copy = ImmutableSet.copyOf(issues);
 
         for (JiraIssue issue : copy) {
-            logger.println(Messages.UpdatingIssue(issue.id));
+            logger.println(Messages.UpdatingIssue(issue.getKey()));
 
             try {
                 session.addComment(
-                        issue.id,
+                        issue.getKey(),
                         createComment(build, useWikiStyleComments, jenkinsRootUrl, recordScmChanges, issue),
                         groupVisibility, roleVisibility
                 );
                 if (!labels.isEmpty()) {
-                    session.addLabels(issue.id, labels);
+                    session.addLabels(issue.getKey(), labels);
                 }
 
             } catch (RestClientException e) {
 
                 if (e.getStatusCode().or(0).equals(404)) {
-                    logger.println(issue.id + " - JIRA issue not found. Dropping comment from update queue.");
+                    logger.println(issue.getKey() + " - JIRA issue not found. Dropping comment from update queue.");
                 }
 
                 if (e.getStatusCode().or(0).equals(403)) {
-                    logger.println(issue.id + " - Jenkins JIRA user does not have permissions to comment on this issue. Preserving comment for future update.");
+                    logger.println(issue.getKey() + " - Jenkins JIRA user does not have permissions to comment on this issue. Preserving comment for future update.");
                     continue;
                 }
 
                 if (e.getStatusCode().or(0).equals(401)) {
-                    logger.println(issue.id + " - Jenkins JIRA authentication problem. Preserving comment for future update.");
+                    logger.println(issue.getKey() + " - Jenkins JIRA authentication problem. Preserving comment for future update.");
                     continue;
                 }
 
-                logger.println(Messages.FailedToUpdateIssueWithCarryOver(issue.id));
+                logger.println(Messages.FailedToUpdateIssueWithCarryOver(issue.getKey()));
                 logger.println(e.getLocalizedMessage());
             }
-
-            // if no exception is thrown during update, remove from the list as succesfully updated
-            issues.remove(issue);
         }
 
     }
 
-    private static List<JiraIssue> getJiraIssues(Set<String> ids, JiraSession session, PrintStream logger) throws RemoteException {
-        List<JiraIssue> issues = new ArrayList<JiraIssue>(ids.size());
+    private static Set<JiraIssue> getJiraIssues(Set<String> ids, JiraSession session, PrintStream logger) throws RemoteException {
+        Set<JiraIssue> issues = Sets.newHashSet();
```

Comment:
Not sure it makes a huge difference given the amount of allocations likely to happen here? Ive restored it anyway.